datafusion/physical-plan/src/windows/bounded_window_agg_exec.rs

```diff
@@ -371,23 +375,25 @@ trait PartitionSearcher: Send {
         window_expr: &[Arc<dyn WindowExpr>],
         partition_buffers: &mut PartitionBatches,
     ) -> Result<()> {
-        if record_batch.num_rows() > 0 {
-            let partition_batches =
-                self.evaluate_partition_batches(&record_batch, window_expr)?;
-            for (partition_row, partition_batch) in partition_batches {
-                let partition_batch_state = partition_buffers
-                    .entry(partition_row)
-                    .or_insert_with(|| PartitionBatchState {
-                        record_batch: RecordBatch::new_empty(partition_batch.schema()),
-                        is_end: false,
-                        n_out_row: 0,
-                    });
-                partition_batch_state.record_batch = concat_batches(
-                    &partition_batch.schema(),
-                    [&partition_batch_state.record_batch, &partition_batch],
-                )?;
+        if record_batch.num_rows() == 0 {
+            return Ok(());
+        }
+        let partition_batches =
+            self.evaluate_partition_batches(&record_batch, window_expr)?;
+        for (partition_row, partition_batch) in partition_batches {
+            let partition_batch_state = partition_buffers
+                .entry(partition_row)
+                .or_insert_with(|| PartitionBatchState::new(partition_batch.schema()));
+            partition_batch_state.extend(&partition_batch)?;
+        }
+
+        if self.is_mode_linear() {
+            let last_row = get_last_row_batch(&record_batch)?;
+            for (_partition_row, partition_batch) in partition_buffers.iter_mut() {
+                partition_batch.set_most_recent_row(last_row.clone());
```

Comment:
This is a global attribute and does not change from partition to partition. Maybe we can find a way to make this global since they are always equal to each other in theory.